src/emlid_reach.cpp

```diff
@@ -0,0 +1,434 @@
+ */
+
+
+#include "emlid_reach.h"
+
+#include <stdio.h>
+#include <math.h>
+#include <string.h>
+#include <ctime>
+#include <cmath>
+#include <cstring>
+#include <errno.h>
+
+#include <stdlib.h>
+
+//// ERB
+// 'E' Ox45 | 'R' Ox52 | ID | LENGTH (2 Bytes little endian) | payload | CHECKSUM (2B)
+#define ERB_SYNC_1 0x45 // E
+#define ERB_SYNC_2 0x52 // R
+
+#define ERB_ID_VERSION         0x01
+#define ERB_ID_GEODIC_POSITION 0x02
+#define ERB_ID_NAV_STATUS      0x03
+#define ERB_ID_DOPS            0x04
+#define ERB_ID_VELOCITY_NED    0x05
+#define ERB_ID_SPACE_INFO      0x06 // not used, reduces stack usage
+#define ERB_ID_RTK             0x07 // RTK, TODO
+
+
+#define EMLID_UNUSED(x) (void)x;
+
+#define GPS_PI 3.141592653589793238462643383280f
+
+#define AUTO_DETECT_MAX_TIMEOUT       10 // if more detect failed AND
+#define AUTO_DETECT_MAX_PARSE_ERR     2  // if more detect failed AND
+#define AUTO_DETECT_MAX_READ_SENTENCE 10 // if more detect succeed
+
+
+GPSDriverEmlidReach::GPSDriverEmlidReach(GPSCallbackPtr callback, void *callback_user,
+		struct vehicle_gps_position_s *gps_position, struct satellite_info_s *satellite_info) :
+	GPSHelper(callback, callback_user),
+	_gps_position(gps_position), _satellite_info(satellite_info)
+{}
+
+
+int
+GPSDriverEmlidReach::configure(unsigned &baudrate, OutputMode output_mode)
+{
+	// TODO RTK
+	if (output_mode != OutputMode::GPS) {
+		GPS_WARN("EMLIDREACH: Unsupported Output Mode %i", (int)output_mode);
+		return -1;
+	}
+
+	unsigned baud_allowed[] {57600, 115200, 230400};
+
+	for (unsigned i = 0; i < sizeof(baud_allowed) / sizeof(baud_allowed[0]); i++) {
+		if (baudrate > 0 && baudrate != baud_allowed[i]) {
+			continue;
+		}
+
+		_sentence_cnt = 0;
+		_parse_err_cnt = 0;
+		_erb_decode_state  = ERB_State::init;
+		_read_buff_len = 0;
+		_checksum_buff_len = 0;
+		_read_buff_ptr = _read_buff + GPS_READ_BUFFER_SIZE;
+		_erb_payload_len = 0;
+
+		if (GPSHelper::setBaudrate(baud_allowed[i]) != 0) {
+			continue;
+		}
+
+		if (! testConnection()) {
+			continue;
+		}
+
+		GPS_INFO("EmlidReach: config OK with ERB baudrate: %d", baud_allowed[i]);
+		baudrate = baud_allowed[i];
+		return 0;
+	}
+
+	return -1;
+}
+
+
+bool
+GPSDriverEmlidReach::testConnection()
+{
+	_testing_connection = true;
+
+	unsigned timeout_cnt = 0;
+
+	while (timeout_cnt < AUTO_DETECT_MAX_TIMEOUT
+	       && _parse_err_cnt < AUTO_DETECT_MAX_PARSE_ERR
+	       && _sentence_cnt < AUTO_DETECT_MAX_READ_SENTENCE) {
+		if (receive(50000) == 0b100) { // timeout larger than what defined in read() are truncated
+			timeout_cnt ++;
+		}
+	}
+
+	_testing_connection = false;
+	return timeout_cnt < AUTO_DETECT_MAX_TIMEOUT && _parse_err_cnt < AUTO_DETECT_MAX_PARSE_ERR;
+}
+
+
+int
+GPSDriverEmlidReach::receive(unsigned timeout)
+{
+	while (true) {
+		if (_read_buff_ptr >= _read_buff + _read_buff_len) {
+			// no more data in buff to process, read from serial
+			_read_buff_len = read(_read_buff, sizeof(_read_buff), timeout);
```

Comment:
Instead of reading into `_read_buff` and later on parsing into `_buff`, why not reading into a local buffer `uint8_t buf[GPS_READ_BUFFER_SIZE];` and then directly call `erbParseChar` on that?

Reply:
GpsHelper::receive() interface returns when there is data to publish. This means that if there is still bytes to be parsed after an ERB (position) sentence is received, those remaining bytes are lost if the buffer is local.
The only way to use a local buff and to be sure there is not data loss, would be that the read() function from the serial device guarantees that data is read line by line. While testing, I could see that data is not always read line by line (even if it appears to reads line by line at first), that is why _read_buff was moved from local scope to class scope

If the aim is to avoid having 2 buffers (_read_buff and _buff), that could be achieved using memmove() and pointer arithmetic

@bkueng what do you say ? Should I go for using 1 buff and memmove()

Reply:
The idea would be:
- read into the local buffer
- parse everything that was received for that (via `erbParseChar`)
- if there was a position message parsed, return, otherwise iterate and check for the timeout.

Which means that there are no bytes lost (a partial message would still be kept in `_buff` and continued in the next `receive` call). What could happen is that a `read` contains already multiple position messages. In that case only the last one would be returned, and it means that `receive` is not called fast enough.
Does that happen to you? What is the position update rate?

Reply:
Position update rate is around 5Hz. Practically, it goes between 4.9 a,d 5.1 Hz. When sat view is not good, it may go down (sky is blocked by building for example) if the lock is not stable

I understand the idea, and the limitation associated to it ("only the last would be returned")
I disagree on "it means that receive is not called fast enough", because several message types are received (status, position, ned, etc), they are received by bursts (all types 5 times per sec), and are not all relevant. That means that a message I don't care about (sat info for example), can shadow useful messages (position, ned).
Since ubx is using an rx_buff and a sentence_buff (packed struct), I propose to do the same here (I don't want to change a winning team)
The footprint of the rx_buff is 150, as defined in GpsHelper 
46:#define GPS_READ_BUFFER_SIZE 150

Secondly, regarding the local vs object scope, since the waiting time is happening when serial.read() is called, that means that most of the lifetime of the GPS is happening within the GpsHelper::receive(), which is the local scope mentioned above. I conclude from that that moving the buff to object scope is of very little side effect. Or am I missing something ?

@bkueng Looking forward for your answer :-)

Reply:
>  because several message types are received (status, position, ned, etc)

You would only return on a position update - since that is what we really care about. Then my argument holds (otherwise you're right, it would not).

> I conclude from that that moving the buff to object scope is of very little side effect.

The side effect is the stack size requirement: since other drivers have it on the stack, they impose a minimum stack size to the gps driver. If the emilid driver moves the buffer to the heap (object scope), that stack space is unused, and therefore overall more memory is used.
Or in other words: what's important is that all drivers do (roughtly) the same thing to minimize overall memory usage. Doing it the other way around for all drivers would work too.

Reply:
@bkueng Thanks for the time to answer
Indeed, I didn't consider heap vs stack, just memory vs time... oops :-)
Ok, so I do it on stack, your argument is good, we don't want to waste stack reserved by other drivers

Regarding 
> You would only return on a position update - since that is what we really care about. Then my argument holds (otherwise you're right, it would not).
> 
That is partially true, because some of the info returned with position are obtained with other messages, namely DOPs, NED and fix status.
I think now I'll make the buff var local (so stack), and I will return (the position) only after read() buffer has been fully parsed. That should meet all requirements